jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -755,29 +755,37 @@ def mos_niriss_parser(app, data_dir):
             print(f"Loading: {flabel} sources")
 
             with fits.open(fname, memmap=False) as temp:
-                # treat all HDUs missing SRCTYPE attribute as extended sources
-                # TODO: Remove this once valid SRCTYPE values are present in all headers
+                filtered_hdul = fits.HDUList()
+
+                # Manually copy over known extensions if they exist
+                for metahdu in ('PRIMARY', 'ASDF'):
+                    if metahdu in temp:
+                        filtered_hdul.append(temp[metahdu])
+
                 for hdu in temp:
-                    if ("SRCTYPE" in hdu.header
-                            and (hdu.header["SRCTYPE"] in ("POINT", "EXTENDED"))):
-                        pass
-                    else:
-                        hdu.header["SRCTYPE"] = "EXTENDED"
+                    # Only load the SOURCEIDs the user specified in the target catalog
+                    if hdu.header.get('SOURCEID', None) in cat_id_dict.keys():
+
+                        # Specutils parser requires SRCTYPE to be either POINT or EXTENDED
+                        # treat all HDUs missing SRCTYPE attribute as extended sources
+                        # TODO: Remove this once valid SRCTYPE values are present in all headers
+                        if hdu.header.get('SRCTYPE', None) not in ("POINT", "EXTENDED"):
+                            hdu.header["SRCTYPE"] = "EXTENDED"
+
+                        filtered_hdul.append(hdu)
```

Comment:
can these two for loops be handled at once?  Or do PRIMARY and ASDF need to be included first?  And if they can be merged - would a list comprehension (perhaps with a function to do the filtering - maybe something like `filtered_hdul = fits.HDUList([hdu for hud in temp if include_hdu(hdu)])`) be cheaper than appending?

Reply:
Agh thanks for bringing this up! I forgot to implement one last improvement that I think addresses this. I changed the metahdu logic to `pop` the hdu rather than just reference it in https://github.com/spacetelescope/jdaviz/pull/1696/commits/1b332f4ed8821e6c5c1586365c568ae54ba3c7ae. `pop` removes the hdus, so effectively, the code is only looping over the full contents of `temp` once. Is this satisfactory, or do you think list comprehension would still be cheaper?

Reply:
If there isn't a need to preserve this specific order, I think list comprehension should be cheaper (and easier to read/maintain, imo).  Might at least be worth a shot to compare?  If you'd rather stick with the loops, though, let me know and I can test and finish reviewing.

Reply:
Thanks to us removing the SRCTYPE hack, I think I was able to get it to work! Looks a lot nicer now; ready for re-review!